soc: nordic: add cache helpers

nika-nordic opened the pull request:
(Simplified PR description from original proposal, since the scope is vastly reduced)

PR adds cache helpers specific to nRF devices that streamline the process of allocating DMA buffer in correct memory region and managing the data cache on nRF54H20.

nika-nordic committed:
soc: nordic: add dmm component

DMM stands for Device Memory Management and its role is to streamline
the process of allocating DMA buffer in correct memory region
and managing the data cache.

Signed-off-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>

nika-nordic commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ * @retval -ENOMEM If output buffer could not be allocated.
+ * @retval -errno Negative errno for other failures.
+ */
+int dmm_dma_buffer_out_prepare(const struct device *dev, void const * user_buffer, void ** buffer_out);
```
it would be good  to discuss whether `{in,out}` function variants should be merged into one with additional parameter like `direction={IN/OUT}`

hubertmis replied:
I don't think so. These functions are used in different contexts, so it is cleaner to keep them with different symbols. It's easier to follow the code this way.

hubertmis commented on soc/nordic/common/CMakeLists.txt:
```diff
@@ -7,6 +7,8 @@ if(CONFIG_ARM AND CONFIG_SOC_FAMILY_NORDIC_NRF)
   zephyr_library_sources(soc_nrf_common.S)
 endif()
 
+zephyr_library_sources(dmm.c)
```
nit: I guess some `IFDEF` is required for this line to avoid compiling dmm.c for targets without data cache

nika-nordic replied:
sure, I will do it properly in final implementation for legacy devices like nRF52

hubertmis commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ *       dynamic allocation and cache operations might be skipped.
+ *
+ * @warning It is prohibited to read or write @p user_buffer or @p buffer_out contents
+ *          until @ref dmm_dma_buffer_out_release() is called
```
```suggestion
 *          from the time this function is called until @ref dmm_dma_buffer_out_release() is called on the same buffer or this function returns with an error
```

nika-nordic replied:
fixed

hubertmis commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ *          until @ref dmm_dma_buffer_out_release() is called
+ *
+ * @param dev Device to prepare the buffer for.
+ * @param user_buffer Buffer containing data to be processed by the given device
```
```suggestion
 * @param user_buffer CPU address (virtual if available) of the buffer containing data to be processed by the given device
```

nika-nordic replied:
fixed

hubertmis commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ *
+ * @param dev Device to prepare the buffer for.
+ * @param user_buffer Buffer containing data to be processed by the given device
+ * @param user_out Pointer to a pointer to be filled by prepared DMA buffer address.
```
```suggestion
 * @param user_out Pointer to a bus address of a buffer containing the prepared DMA buffer
```

nika-nordic replied:
fixed

hubertmis commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ * @brief Release the previously prepared DMA output buffer
+ *
+ * @param dev Device to release the buffer for.
+ * @param buffer_out DMA output buffer previously prepared with @ref dmm_dma_buffer_out_prepare().
```
```suggestion
 * @param buffer_out Bus address of the DMA output buffer previously prepared with @ref dmm_dma_buffer_out_prepare().
```

nika-nordic replied:
fixed

hubertmis commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ *       dynamic allocation might be skipped.
+ *
+ * @warning It is prohibited to read or write @p user_buffer or @p buffer_in contents
+ *          until @ref dmm_dma_buffer_in_release() is called
```
ditto, we should be explicit when the time period of not-accessing starts (call or return of the function?) and does it ends when function returns an error

hubertmis commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ *          until @ref dmm_dma_buffer_in_release() is called
+ *
+ * @param dev Device to prepare the buffer for.
+ * @param user_buffer Buffer to be filled with data from the given device
```
```suggestion
 * @param user_buffer CPU address (virtual if available) of the buffer to be filled with data from the given device
```

nika-nordic replied:
fixed

hubertmis commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ *
+ * @param dev Device to prepare the buffer for.
+ * @param user_buffer Buffer to be filled with data from the given device
+ * @param user_in Pointer to a pointer to be filled by prepared DMA buffer address.
```
```suggestion
 * @param user_in Pointer to a bus address of the buffer to be filled by DMA
```

nika-nordic replied:
fixed

hubertmis commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ * Copy @p buffer_in contents into @p user_buffer, if needed.
+ *
+ * @param dev Device to release the buffer for.
+ * @param user_buffer Buffer to be filled with data from the given device
```
```suggestion
 * @param user_buffer CPU address of a buffer to be filled with data from the given device
```

nika-nordic replied:
fixed

hubertmis commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ *
+ * @param dev Device to release the buffer for.
+ * @param user_buffer Buffer to be filled with data from the given device
+ * @param buffer_in DMA input buffer previously prepared with @ref dmm_dma_buffer_in_prepare().
```
```suggestion
 * @param buffer_in Bus address of the DMA input buffer previously prepared with @ref dmm_dma_buffer_in_prepare().
```

nika-nordic replied:
fixed

hubertmis commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ *
+ * Invalidate data cache lines associated with input buffer, if needed.
+ * Copy @p buffer_in contents into @p user_buffer, if needed.
+ *
```
I would add a note that buffer_in and user_buffer parameters pair must have the same values as in the matching call to `dmm_dma_buffer_in_prepare()`

hubertmis commented on soc/nordic/common/dmm.c:
```diff
@@ -0,0 +1,58 @@
+    // Get memory region that specified device can perform DMA transfers from, using devicetree
+
+    // Check if provided user buffer is already in correct memory region
+    // If yes, assign buffer_out to user_buffer
```
First, check if the buffer is aligned and padded to cache lines, or if it is in non-cacheable memory. Only if these conditions are met assign it to user_buffer.

nika-nordic replied:
added

hubertmis commented on soc/nordic/common/dmm.c:
```diff
@@ -0,0 +1,58 @@
+
+int dmm_dma_buffer_out_release(void const * buffer_out)
+{
+    // Get memory region that specified device can perform DMA transfers from, using devicetree
```
Is this step needed here?

hubertmis replied:
So the function should get the memory region controlled by the dynamic allocator, which is not necessarily the same as the one from which DMA can perform a transfer. If the API user passes a static DMAable buffer it must not be freed in this function.

I asked if this action was needed because I assumed the check in the next line implicitly gets the memory range supervised by the allocator to compare the buffer address with the dynamic allocator range.

nika-nordic replied:
so we are missing `dev` parameter - once this is available we can perform the action you described

hubertmis replied:
Yes, I'm not sure it's required in all cases, but I can imagine scenarios in which it might.
Also, adding a `dev` parameter is good for function parameters symmetry.

hubertmis commented on soc/nordic/common/dmm.c:
```diff
@@ -0,0 +1,58 @@
+{
+    // Get memory region that specified device can perform DMA transfers to, using devicetree
+
+    // Check if provided user buffer is already in correct memory region
```
```suggestion
    // Check if provided user buffer is already in correct memory region. If this region is cacheable, check if the buffer is aligned and padded to the cache lines.
```

nika-nordic replied:
fixed

hubertmis commented on soc/nordic/common/dmm.c:
```diff
@@ -0,0 +1,58 @@
+    // Check if provided user buffer is already in correct memory region
+    // If yes, assign buffer_in to user_buffer
+    // If no, dynamically allocate buffer in correct memory region, using for example mem_attr_heap_alloc()
+
```
If the buffer is cacheable, invalidate it to clear potential dirty bits in the cache.

nika-nordic replied:
why is it needed? As far as I know DMA does not respect cache line status before writing to memory, and the cache lines will be invalidated during `dmm_dma_buffer_in_release` call (which is to be invoked once the DMA fills the buffer)

hubertmis replied:
Let's assume a cache line is tagged for the *buffer_in, the cache line contains the value 0xdeadbeef and the whole line is marked as dirty. The *buffer_in in RAM contains 0x00000000. After this function is called, the DMA buffer starts. After 2 bytes are received: 0x12340000 another software thread reuses the same line in cache to load data from another address. This triggers a hardware write-back of the dirty line. The write-back overwrites the memory at *buffer_in to 0xdeadbeef. The DMA transfer progresses and when it finishes the *buffer_in in RAM contains 0xdead5678 instead of 0x12345678. This is a data corruption scenario that must be prevented.

To prevent the data corruption scenario I described, the potential cache line tagged for the input DMA buffer cannot have any dirty bits set. The most efficient way to clean all dirty bits is invalidating a cache line.

nika-nordic replied:
isn't this scenario avoided if we allocate buffers at cache line boundary and pad buffer size to cache line size?

hubertmis replied:
Let me expand my scenario:

The DMA buffer is provided by the driver user (application) with code like this:

```
uint8_t buffer[4] ALIGNED(CACHE_LINE) PADDED(CACHE_LINE) IN_DMABLE_MEMORY_FOR(spi0);

*(uint32_t *)buffer = 0xdeadbeef;

spi_receive(&spi0, buffer, ...);
```

So the buffer is at the cache line boundary. Yet when the driver starts execution, the cache line can contain dirty data `0xdeadbeef`. To make sure this dirty line does not cause data corruption, it must be invalidated before DMA starts writing to the buffer.

nika-nordic replied:
> To make sure this dirty line does not cause data corruption

How this dirty line can cause data corruption? When cache line is written-back after DMA finalizes the RX transfer but before cache line invalidation?

hubertmis replied:
> > To make sure this dirty line does not cause data corruption
> 
> How this dirty line can cause data corruption? When cache line is written-back after DMA finalizes the RX transfer but before cache line invalidation on RX event handling?

Yes. To be precise the data corruption happens if the HW initializes write-back automatically (cache line eviction) during a time window. The time window starts when the first data unit gets stored by DMA in the buffer, and it ends when the driver invalidates the line.

nika-nordic replied:
added

hubertmis commented on soc/nordic/common/dmm.c:
```diff
@@ -0,0 +1,58 @@
+    // Get memory region that specified device can perform DMA transfers to, using devicetree
+
+    // Check if input buffer is contained within memory area managed by dynamic memory allocator
+    // If yes, free the buffer
```
I wouldn't free the buffer until its content is copied to the user buffer.

nika-nordic replied:
fixed - changed the order of the operations

hubertmis reviewed (changes requested):
I suggested clarifications in API description.

But the issue requiring changes is the algorithm description in dmm.c. Current one could fail in some scenarios (race conditions).

nordic-krch commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ * @retval -ENOMEM If output buffer could not be allocated.
+ * @retval -errno Negative errno for other failures.
+ */
+int dmm_dma_buffer_out_prepare(const struct device *dev, void const * user_buffer, void ** buffer_out);
```
don't you need to provide length?

nika-nordic replied:
length added to `{in/out}_prepare` and `in_release` to make sure allocation/copying knows how much data there is

nordic-krch commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ * @retval -ENOMEM If input buffer could not be allocated.
+ * @retval -errno Negative errno for other failures.
+ */
+int dmm_dma_buffer_in_prepare(const struct device *dev, void const * user_buffer, void ** buffer_in);
```
since in `dmm_dma_buffer_in_release` you are also providing `user_buffer` then it seems that `user_buffer` is redundant here.

nika-nordic replied:
the thing is that `dmm_dma_buffer_in_prepare` needs to check whether the provided buffer has correct attributes (location, alignment) before performing dynamic allocation:
- if `user_buffer` is in correct location, reachable by DMA of given device, and has correct attributes, then `*buffer_in = user_buffer;`
- otherwise `*buffer_in = dynamically_allocate(user_length);`

nordic-krch commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+ * @retval -ENOMEM If input buffer could not be allocated.
+ * @retval -errno Negative errno for other failures.
+ */
+int dmm_dma_buffer_in_prepare(const struct device *dev, void const * user_buffer, void ** buffer_in);
```
I wonder if this function should not be split into `alloc` and `prepare`. Currently this api does not cover case when driver would like to allocated at compile time a buffer (e.g UART for polling out or for RX flushing) and then use this API only for cache management (if we want to use only dmm for cache management). `alloc` might be common for both directions.

For dynamic buffer it would look like
```
err = dmm_dma_buffer_alloc(dev, &dma_buffer, length);
if (err == 0) dmm_dma_buffer_out_prepare(dev, dma_buffer, user_buffer, length);
```
For statically allocated buffers
```
uint8_t dma_buffer[length] MEM_REGION(...);

// Prepare function just copies and performs cache writeback (if needed)
dmm_dma_buffer_out_prepare(dev, dma_buffer, user_buffer, length2); //length2 <= length
```

hubertmis replied:
As a next step dmm API should be extended with macros for static buffer allocation. These macros can be used by the drivers covering the scenario you described, or by the applications to improve the performance of the drivers.

If the buffers are allocated using the to-be-created dmm macros, they are guaranteed to be bypassed by `dmm_dma_buffer_in/out_prepare()` without dynamic buffer allocation and copying overhead.

I don't think the driver should have any indication if the buffer in use is to be bypassed or not by dmm. The driver should not track the qualities of the buffers, because it's dmm's responsibility. The driver should blindly call `dmm_..._prepare()` on any buffer.

nika-nordic replied:
> Currently this api does not cover case when driver would like to allocated at compile time a buffer (e.g UART for polling out or for RX flushing) and then use this API only for cache management (if we want to use only dmm for cache management). 

Actually this API covers that - if the buffer has correct attributes (location and cache alignment) then no dynamic allocation will happen and `dmm` will simply return what was provided to it

nordic-krch commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,96 @@
+/**
+ * @brief Release the previously prepared DMA output buffer
+ *
+ * @param dev Device to release the buffer for.
```
Is this the device that will use DMA (e.g. UARTE) or mempool device that is associated with the DMA device through DT reference?
Imo, it cannot be the peripheral because there is no generic, runtime device API to get mempool reference from the device.

hubertmis replied:
No. The device drivers must be generic: they must work correctly with the compatible peripherals regardless of how they are integrated into the system (which memory is DMAble by them, what are properties of this memory). The integration information should be included in the device tree properties of the compatible node and retrieved by the device driver. The device driver should use only the node it instantiates.

If a device tree node misses a property like `mempool`, then maybe the driver should not use `dmm` at all.

hubertmis replied:
Scratch my previous message. I misunderstood the proposal.

A device driver X uses a reference to a device tree node it instantiates (N), with no changes compared to the current device drivers.
What we would modify is to add a property M to node N being a reference to a mempool. The device driver X would retrieve property M from node N and use `const struct dev*` (D) represented by M. When X calls `dmm` functions it would use D as the `dev` argument.

If I correctly described the proposal, then it's feasible.

nika-nordic replied:
I see another issue here. In `dmm.c` `_prepare` API we need to check whether the provided buffer is already in correct memory region and if not, use mempool device to dynamically allocate new buffer.

Does it mean we actually need to provide **two** devices to `_prepare` API:
- reference to RAM region node, associated with specified peripheral node (to check buffer location),
- reference to mempool device (to perform dynamic allocation if buffer is in incorrect location) ?

If yes, I am wondering if these two could be somehow linked, so we provide single device instead of two.

> Imo, it cannot be the peripheral because there is no generic, runtime device API to get mempool reference from the device.
@nordic-krch would be it feasible to add such runtime device API? I believe it would improve the abstraction and also solve this issue

hubertmis replied:
> Does it mean we actually need to provide **two** devices to `_prepare` API

Good point. I don't think a device tree node should have an indication of software configuration like if a given memory region contains a pool for dynamically allocated memory. The device tree node should describe the hardware properties like which memory region is DMA-able by this peripheral. Something like:

```
ram30: memory@0x... {
  properties = NON_CACHEABLE;
};

uart130 {
  compatible = "nordic,...";
  memory = <&ram30 &ram31>;
};
```

And dmm should figure out internally which allocator to use to correctly allocate a buffer in `memory` pointed by `&uart130`.

nika-nordic replied:
updated PR - now `dmm` accepts memory region address pointer as first parameter. Device drivers (and other components) can easily obtain this value by peeking at devicetree property. `dmm` has internal list of regions and finds the one matching the one provided by the user of the API

nordic-krch commented:
I've been experimenting with cache, mvdma and access time to RAM (in particular RAM3 used for slow peripherals). I've noticed that access to RAM3 is very slow. On average it takes ~0.37us to transfer a byte from global RAM (cacheable) to RAM3 and ~0.3us to transfer from RAM3 to global RAM. Dependency is linear so memcpy of 100 bytes takes 30-37us (100 cycles per byte, all with DCACHE enabled). It is a lot of time. On the other hand, setting up a MVDMA transfer takes ~4us (it's mainly writing back data and job descriptors from DCACHE). MVDMA transfer takes approximately same time as memcpy. Given that, I think that it might be good to have asynchronous API as well. Something like:

```
void handler(void *dmm_buffer, size_t length, void *user_data)
{
    struct device *dev = (struct device *)user_data;
    buffer_set(dev, dmm_buffer, length);
    transfer_start(dev);
}

dmm_buffer_prepare(void *buffer, size_t len, dmm_handler_t handler, void *user_data)
```
Depending on the threshold user handler would be called directly from the context of `buffer_prepare` call or from MVDMA interrupt handler.

nika-nordic commented:
@nordic-krch alternatively we could make the `_prepare` call always blocking and handle MVDMA in `dmm` context, but I guess the consequence is that `_prepare` API will no longer be usable in IRQ context and it needs to be (for example for UARTE ENDRX event), correct?

My understanding is that adding `handler` parameter to both `out_prepare` and `in_prepare` will make the code more flexible:
- in first version of the code, without MVDMA, we can simply call `handler` at the end of the `_prepare` function execution
- once the MVDMA handling is added, `handler` could be called asynchronously from MVDMA interrupt context
- if user wants to enforce copying in-place then handler can be NULL
Let me know if it makes sense (cc: @hubertmis as this is extension of original design )

nika-nordic commented on soc/nordic/common/dmm.h:
```diff
@@ -0,0 +1,107 @@
+ * @retval -ENOMEM If output buffer could not be allocated.
+ * @retval -errno Negative errno for other failures.
+ */
+int dmm_dma_buffer_out_prepare(const struct device *dev, void const * user_buffer, size_t user_length, void ** buffer_out);
```
general comment - now I see that this `_dma_` part could be removed from all of the APIs, as it feels to be redundant

hubertmis commented:
> * cc: @hubertmis as this is extension of original design

I don't know the details of the requirement to manage DMA buffers directly from ISRs. I would expect ISRs to be short and avoid `memcpy()` or setting DMA transfers in there. Why not simply defer the execution from an ISR context to a thread context (or a workqueue) and call blocking DMA functions from there?

nordic-krch commented:
We already have performance issues due to peripheral register access and buffer management so imo having an option to use DMA and reduce CPU is a plus even if there will be cases when it is not used. Deferring to thread context takes additional amount of time (posting a semaphore is few us) and we already have asynchronous API in UART where next buffer is requested from the interrupt context and there is rtio allows asynchronous i2c and spi transfers (not yet supported by nordic).

> for MVDMA the buffers must be prepared as well, potentially copied to the correctly aligned and padded cacheable buffers.

I think that at some point zephyr will get `__cacheline_aligned` attribute (like in linux https://elixir.bootlin.com/linux/latest/source/include/linux/cache.h#L53) and buffers used for drivers will be tagged with that (e.g. in sensor drivers) so dma-dmm might be used then. `__cacheline_aligned` would use dedicated section so it is possible to detect if address is from that section and act accordingly. Something like:
```
int dmm_prepare(buf, len, handler, user_data)
{
  if (in_cache_section(buf) && len > thr) {
     dma_dmm(buf, len, handler, user_data);
} else {
    dmm_buf = memcpy_dmm(buf, len);
    handler(dmm_buf, len, user_data);
}
```

> we would get a recursive execution

Sorry, I didn't get where would recursive execution would happen.

nika-nordic commented:
> think that at some point zephyr will get __cacheline_aligned attribute (like in linux https://elixir.bootlin.com/linux/latest/source/include/linux/cache.h#L53) and buffers used for drivers will be tagged with that (e.g. in sensor drivers) so dma-dmm might be used then. __cacheline_aligned would use dedicated section so it is possible to detect if address is from that section and act accordingly.

Please take a look at the pseudo-code for `_prepare` API in `dmm.c` - when provided buffer is in correct memory region and is correctly aligned then the provided buffer is simply returned to the user, without dynamic allocation. According to original design this logic would be present in `dmm.c` so the driver does not need to contain any additional logic.

> Sorry, I didn't get where would recursive execution would happen.

If MVDMA driver would use `dmm` as well then the recursive execution could happen. So MVDMA driver in this use-case would need to use buffers allocated at compile-time to avoid dynamic allocation and extra copying.

> (...) and we already have asynchronous API in UART where next buffer is requested from the interrupt context and there is rtio allows asynchronous i2c and spi transfers (not yet supported by nordic).

I think simply adding `handler` as additional parameter to `_prepare` API would allow us to cover this case in the future, as I described in: https://github.com/zephyrproject-rtos/zephyr/pull/71278#issuecomment-2057116351

hubertmis commented:
> I think that at some point zephyr will get `__cacheline_aligned` attribute

I don't think this will happen due to performance reasons.

Let's discuss an example:

We can get a network frame through Bluetooth or Wi-Fi radio. The network frame is stored in network buffers. A part of the network frame structure is payload which we could want to transmit using I2S, UART, or whatever. This payload in network buffers is not aligned to anything DMA-related.

In Linux, a user-space application would take the network frame, parse it, and send it to I2S or UART driver. This would require data to cross kernel/user space two times, copying the buffers, and during each copy, the buffer could get aligned with `__cacheline_aligned`.

But in Zephyr in MCUs, we prefer a zero-copy model, so if that's possible I2S should DMA data out directly from the network buffers. Only in architectures in which that's not possible, the data should be copied.

I can imagine application code like this:

```
net_buf_len = recv(socket, net_buffer);
result = parse_net_frame(net_buffer, net_buf_len, &payload, &payload_len); // get pointer to the middle of the network frame
if (result) {
  i2s_send(payload, payload_len);
}
```

As you can see, nothing in this code gets aligned by the application. The network buffer is aligned to the Bluetooth or Wi-Fi's DMA. The payload is potentially moved and aligned to I2S DMA by the I2S driver. This data movement would happen in nRF54H, but never in nRF52 or nRF54L.

But I agree we can add the `handler` parameter to potentially implement MVDMA helper in the future. Maybe something like copying the unaligned ends of the buffer with `memcpy` and the middle part by MVDMA.

nordic-krch commented:
> I don't think this will happen due to performance reasons.

That's the example where it might e.g. sensor, shell transport having a static buffer for transfering data can be tagged.

hubertmis commented:
> > I don't think this will happen due to performance reasons.
> 
> That's the example where it might e.g. sensor, shell transport having a static buffer for transfering data can be tagged.

Sure, there are examples in which buffers can be tagged by the application. There is an idea to extend the dmm API with such tags. But we need to be prepared for tag-less buffers as well because there are also scenarios in which tags won't be used.

nika-nordic commented:
@hubertmis @nordic-krch
Can you describe what `buffer tagging` is for future reference?

hubertmis commented:
> @hubertmis @nordic-krch Can you describe what `buffer tagging` is for future reference?

I understood tagging as adding preprocessor macros ensuring DMAbility of the buffers like:

```
uint8_t untagged_buffer[BUFFER_SIZE];
uint8_t tagged_buffer[BUFFER_SIZE] DMM_MEM_DMABLE_BY(device_tree_node);
```

The tag would expand to setting a linker section containing RAM achievable by DMA, and to alignment/padding parameters if caching requires them.

kartben commented on samples/boards/nrf/dmm/src/main.c:
could you also add a README for this sample?

hubertmis commented on soc/nordic/common/dmm.c:
```diff
@@ -0,0 +1,272 @@
+	size_t buffer_end = addr + user_length;
+
+	/* Check if the buffer is entirely contained in the region */
+	if (dmm_is_contained_in_region(addr, user_length, region->dt_addr, region->dt_size)) {
+		/* Check if the buffer is in cacheable region */
+		if (dmm_is_cacheable(region)) {
+			/* Check if the buffer is aligned and padded to the cache line */
+			if (IS_ALIGNED(addr, DMM_DCACHE_LINE_SIZE) &&
+			    IS_ALIGNED(buffer_end, DMM_DCACHE_LINE_SIZE)) {
+				return true;
+			}
+			return false;
+		}
+		return true;
+	}
+	return false;
```
I would refactor this tree to make it easier to follow the logic:

```
if (!dmm_is_buffer_in_region(...)) {
  return false;
}

if (!dmm_is_region_cacheable(...)) {
  // The user-provided buffer is in a correct, uncacheable region. Use it as it is
  return true;
}

if (buffer_is_cache_aligned_and_padded(...)) {
  return true;
}

return false;
```

hubertmis commented on soc/nordic/common/dmm.c:
```diff
@@ -0,0 +1,272 @@
+	return (dh->region->dt_size - (dmm_heap_start_get(dh) - dh->region->dt_addr));
+}
+
+static bool dmm_user_buffer_alloc_check(void *user_buffer, size_t user_length,
```
The name of this function is unclear. I would rename to something like `buffer_is_dmable(void *buf, size_t len, const struct dmm_region *dmable_region)`;

hubertmis commented on soc/nordic/common/dmm.c:
```diff
@@ -0,0 +1,272 @@
+	if (!dh) {
+		return -EINVAL;
+	}
+
```
I would validate the arguments here:
* `buffer_in` must be in `region`,
* if the region is cacheable `buffer_in` must be aligned

This validation can be removed in release builds for performance reasons, like with `#ifndef NDEBUG`

hubertmis commented on samples/boards/nrf/dmm/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(hello_world)
```
I'm not sure if this should be a sample or a test. I don't think anyone would use `dmm` from an application code. It is used only from the drivers' code. So a sample showing how to use it from the application code seems to clutter the sample's folder.

hubertmis reviewed (commented):
I like the current shape of this subsystem.
My main concerns are:
* .dtsi modification
*  whether the created sample really should be a sample.

Others are just nit-picking trying to make the code as clean as possible.

frkv commented:
Generic question/comment with regards to "heap" or "heap like" usage. Would it be possible to handle the unaligned data in the stack instead?

We have uses `alloca` (looks like heap but the memory is just placed on stack) and stack size checks for crypto operations in the nRF5 SDK (processed by code logic and finally DMA operations in CryptoCell). 

I understand that this would only be usable for "blocking calls". But for the nRF54H20 and crypto IPC there is (currently) no alternate thing to do than wait for a response (we don't have any async methodology). At most you can use WFI/WFE in the local domain while waiting for the response to come back, but you continue at the same calling context...

I ask this as I see that it could potentially improve the execution, as well as removing some complexities (having a heap like implementation and needing to loop through and lookup items etc)...

nika-nordic commented:
> I ask this as I see that it could potentially improve the execution, as well as removing some complexities (having a heap like implementation and needing to loop through and lookup items etc)...

@frkv Thanks for your input! In general your comment concerns support for different allocation backends, is my impression correct? If yes, I would say it is an performance optimization rather than functional improvement, so maybe we could handle it in another PR? I think right now we should proceed with heap backend, as it is most versatile. If performance is a concern, there is always an option to statically allocate buffers beforehand in appropriate memory region - see SPI driver for reference: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/spi/spi_nrfx_spim.c#L685

frkv commented:
> > I ask this as I see that it could potentially improve the execution, as well as removing some complexities (having a heap like implementation and needing to loop through and lookup items etc)...
> 
> @frkv Thanks for your input! In general your comment concerns support for different allocation backends, is my impression correct? If yes, I would say it is an performance optimization rather than functional improvement, so maybe we could handle it in another PR? I think right now we should proceed with heap backend, as it is most versatile. If performance is a concern, there is always an option to statically allocate buffers beforehand in appropriate memory region - see SPI driver for reference: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/spi/spi_nrfx_spim.c#L685

Absolutely this is follow-up work. It is to the core an optimization. I will bring the attention on the static allocation to the @osaether who is working on auto-generating the IPC frontend. Likely there is an option to have a pattern to avoid lookups to increase performance....

osaether commented on soc/nordic/common/dmm.c:
```diff
@@ -0,0 +1,272 @@
+	// region.
+	if (dmm_user_buffer_alloc_check(user_buffer, user_length, dh->region)) {
+		// If yes, assign buffer_out to user_buffer
+		*buffer_out = user_buffer;
```
In PSA many of the user buffers are const and read only so this code creates a problem.

osaether commented:
It would have been nice with an in/out prepare functions as well. Something like this:
`int dmm_buffer_inout_prepare(void *region, void *user_buffer, size_t user_length, void **buffer)
{
	int err;

	err = dmm_buffer_in_prepare(region, user_buffer, user_length, buffer);
	if (err != 0) {
		return err;
	}
	memcpy(*buffer, user_buffer, user_length);
	return 0;
}`

nika-nordic committed:
tests: boards: nrf: add tests for dmm component

Added tests verify output and input buffers allocation
using dmm component.

Signed-off-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>

nika-nordic committed:
boards: nordic: nrf54h20dk: add DMA attribute to RAM21 & RAM3x

This attribute denotes that DMA operation
can be performed from a given region.

Signed-off-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>

hubertmis commented on drivers/spi/spi_nrfx_spis.c:
```diff
@@ -124,6 +131,22 @@ static int prepare_for_transfer(const struct device *dev,
+
+	err = dmm_buffer_in_prepare(dev_config->mem_reg, rx_buf, rx_buf_len, (void **)&rx_buf);
+	if (err != 0) {
+		LOG_INF("tx alloc err=%d", err);
```
Isn't this a memory leak? Shouldn't you release the `out` buffer here?

hubertmis replied:
And both buffers before `return -EIO` further in this function

hubertmis commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -399,16 +403,15 @@ static void transfer_next_chunk(const struct device *dev)
 		}
 	}
 
+transfer_next_chunk_end:
```
I'm missing where dmm buffers are released in this scenario

hubertmis reviewed (dismissed):
Spotted potential memory leaks in the spi driver. Other than that looks good!

nordic-krch commented on soc/nordic/common/dmm.c:
```diff
@@ -0,0 +1,297 @@
+	return 0;
+}
+
+SYS_INIT(dmm_init, POST_KERNEL, 0);
```
I would move it to PRE_KERNEL. There are cases where we might want to use it earlier (like UART or some crypto stuff). `sys_heap` is a sw component that does not require kernel.

nordic-krch commented on soc/nordic/common/dmm.c:
```diff
@@ -0,0 +1,297 @@
+	if (is_buffer_within_region(addr, 0, dmm_heap_start_get(dh), dmm_heap_size_get(dh))) {
+		/* If yes, free the buffer */
+		dmm_buffer_free(dh, buffer_in);
+	}
```
if buffer is not user buffer and not in the region that should result in error.

nordic-krch replied:
then maybe ` else if (user_length != 0) return -EINVAL;`

carlescufi replied:
Future PR

nordic-krch reviewed (approved):
minor comments which are not blocking.
